### ISSUE TITLE
Transactions fail with Amazon Redshift and RPostgres

### ISSUE THREAD

**Issue by mmuurr** [r-dbi/RPostgres#150]
SQL statements executed within transaction blocks work with Amazon Redshift and the RPostgreSQL DBI-driver package, but are failing with RPostgres as the driver:
```R                                                                                                                       
library(DBI)                                                                                                           
conn1 <- dbConnect(RPostgreSQL::PostgreSQL(), redacted, connection, args)
conn2 <- dbConnect(RPostgres::Postgres(), redacted, connection, args)    
                                                                                                                       
f1 <- function(conn) {                                                                                                 
on.exit(dbRollback(conn), TRUE)                                                                                        
dbBegin(conn)                                                                                                          
dbGetQuery(conn, "SELECT 1")                                                                                           
}                                                                                                                      
                                                                                                                       
f2 <- function(conn) {                                                                                                 
on.exit(dbExecute(conn, "ROLLBACK"))                                                                                   
dbExecute(conn, "BEGIN")                                                                                               
dbGetQuery(conn, "SELECT 1")                                                                                           
}                                                                                                                      
                                                                                                                       
## make sure stand-alone statements work with both drivers:                                                            
dbGetQuery(conn1, "SELECT 1")                                                                                          
#>   ?column?
#> 1        1
dbGetQuery(conn2, "SELECT 1")                                                                                          
#>   ?column?
#> 1        1
                                                                                                                       
## now test SELECT 1 within a transaction:                                                                             
f1(conn1)                                                                                                              
#>   ?column?
#> 1        1
f1(conn2)                                                                                                              
#> Error in result_create(conn@ptr, statement): Failed to prepare query: ERROR:  Query cancelled on user's request
                                                                                                                       
## what about when the transaction is created explicitly with dbExecute?:                                              
f2(conn1)                                                                                                              
#>   ?column?
#> 1        1
f2(conn2)                                                                                                              
#> Error in result_create(conn@ptr, statement): Failed to prepare query: ERROR:  Query cancelled on user's request
```
DBI/driver info:
```R
packageVersion("DBI")        
#> [1] '0.7.13'
packageVersion("RPostgres")  
#> [1] '0.1.12'
packageVersion("RPostgreSQL")
#> [1] '0.6.2'
```
I've seen this problem using both Linux and Mac clients, with these libpq versions:
 * (Amazon) Linux: libpq 9.5.8-1
 * Mac OS 10.11: libpq 9.6.5

I've written [this basic C program](https://gist.github.com/mmuurr/f909ec92182c48e1c026f8901678b13e) to test prepared statements within transactions. The relevant bits are:
```c
  // BEGIN
  res = PQexec(conn, "BEGIN");
  if(PQresultStatus(res) != PGRES_COMMAND_OK) {
    fprintf(stderr, "%s", PQerrorMessage(conn));
    PQclear(res);
    error_exit(conn);
  } else {
    PQclear(res);
  }

  // SELECT (within transaction block)
  res = PQprepare(conn, "", "SELECT 1", 0, NULL);
  if(PQresultStatus(res) != PGRES_COMMAND_OK) {
    fprintf(stderr, "%s", PQerrorMessage(conn));
    PQclear(res);
    error_exit(conn);
  } else {
    fprintf(stdout, "PQresultStatus = %s\n", PQresStatus(PQresultStatus(res)));
    fprintf(stdout, "PQcmdStatus = %s\n", PQcmdStatus(res));
    fprintf(stdout, "PQcmdTuples = %s\n", PQcmdTuples(res));
    PQclear(res);
  }

  res = PQexec(conn, "ROLLBACK");
  PQclear(res);
```
Running the program against both Redshift and PostgreSQL connections yield the same (OK) results:
```bash
PQprotocolVersion = 3
PQresultStatus = PGRES_COMMAND_OK
PQcmdStatus =
PQcmdTuples =
```

Let me know how I can help testing, as I realize not everyone has (the luxury of having :-) an Amazon Redshift cluster available for use.

**Comment by krlmlr** [comment]
Thanks. Does this mean that your C program fails when running against Amazon Redshift when using `PQprepare()` instead of `PQexec()`?

It's fairly easy to use `PQexec()` in the implementation of `dbBegin()` et al. if this solves the problem.

**Comment by mmuurr** [comment]
I think(?) I found the problem.

When `dbBegin()` is called, the single connection object has its `pCurrentResult_` set (to the result of the `BEGIN` command). When subsequent queries are then made, the code path eventually finds its way to [`DbConnection::set_current_result`](https://github.com/r-dbi/RPostgres/blob/ef0ec42a104f852d7e8b6fbbd4d171d38c661fd1/src/DbConnection.cpp#L46), which looks like this:
```c++
void DbConnection::set_current_result(const DbResult* pResult) {
  // Cancels previous query, if needed.
  if (pResult == pCurrentResult_)
    return;

  if (pCurrentResult_ != NULL) {
    if (pResult != NULL)
      warning("Cancelling previous query");

    cleanup_query();
  }
  pCurrentResult_ = pResult;
}
```
Since `pCurrentResult_` is non-null (after the `dbBegin()` call), `cleanup_query()` is called, which in turn calls `finish_query()` **and importantly: `cancel_query()`**.

`finish_query()` plays nicely and iteratively calls `PQgetResult()` until it returns NULL (as described in the libpq docs).

`cancel_query()`, on the other hand, actually (tries) to cancels the query (via `PQcancel()`, which Amazon Redshift interprets as cancelling the transaction. (I'm actually somewhat surprised pure PostgreSQL back-ends doesn't also cancel the transaction here, though their docs are a bit opaque as to how the system should handle a mid-transaction cancellation.)

May I ask why `cancel_query()` is needed in the `cleanup_query()` process? While RPostgres does use libpq's asynchronous patterns, DBI itself seems to operate in synchronous mode, so issuing asynchronous cancellations to the back-end server seems like it introduces trickiness that might not be needed here.

The suggestion of letting `dbBegin()`, `dbCommit()` and `dbRollback()` use `PQexec()` doesn't really solve the problem, since as a transaction proceeds (past the first the `BEGIN`) any subsequent queries will still issue the `PQcancel()` call, aborting the transaction.

For now, I've simply commented-out the `cancel_query()` call in `cleanup_query()` (in [DbConnection.cpp](https://github.com/r-dbi/RPostgres/blob/ef0ec42a104f852d7e8b6fbbd4d171d38c661fd1/src/DbConnection.cpp#L220)), and this resolves the problem.
I'm not 100% sure, however, that the RPostgres asynchronous architecture doesn't need the `PQcancel()` call for some other reason/edge-case ... after all, there's probably a reason you introduced it in the first place?

**Comment by krlmlr** [comment]
@hadley: Do you remember why we need to cancel the query before working with a new result? 5ac028f3

**Comment by krlmlr** [comment]
I think the reason might be to relieve the server of the duty to finish execution of the query. In the case of Amazon RedShift this also seems to cancel a running transaction.

Some queries can return the first rows even before the entire query has been processed. I suspect that even synchronous queries might return to the client before all rows have been processed, but I don't currently have a database large enough where I could investigate this.

**Comment by hadley** [comment]
I don’t remember. It’s quite possible I just got the logic wrong

**Comment by mmuurr** [comment]
The [libpq docs](https://www.postgresql.org/docs/9.6/static/libpq-async.html) state this:

> A client that uses PQsendQuery/PQgetResult can also attempt to cancel a command that is still being processed by the server; see Section 32.6. But regardless of the return value of PQcancel, the application must continue with the normal result-reading sequence using PQgetResult. A successful cancellation will simply cause the command to terminate sooner than it would have otherwise.

The only scenario I can think of in RPostgres where this applies is if `dbClearResult()` is called prior to returning all the rows. In that case, I can see trying to cancel the (possibly-still-running) query prior to `finish_query()`. But if the query is running on the back-end, then the connection's current result (`pCurrentResult_`) should not yet be `complete()`. So in `cleanup_query()`, maybe just performing a check on `pCurrentResult_` for:
 1. presence (i.e. non-NULL) and
 2. completeness?

If present and not-complete, then try to cancel the query (followed by `finish_query()`, which is still needed). And if NULL or complete, then only `finish_query()`. Does that sound close to retaining the original intents of the cancellation procedure?

The cancellation shouldn't ever happen then for `dbBegin()`, since it completes immediately without the need for any `fetch()` calls.

**Comment by mmuurr** [comment]
@krlmlr @hadley, see PR #159 as a potential fix for this issue.
I've verified it solves the transaction problem with Amazon Redshift, and I think it preserves the original intention of `cancel_query()` being part of `cleanup_query()`.
There's likely a more elegant solution here, but I tried to solve this problem without changing any high-level code architecture in RPostgres.

Let me know if I can help any more with this, and thanks much for all the great work!
